src/bucket/BucketApplicator.cpp

```diff
@@ -25,32 +28,38 @@ BucketApplicator::operator bool() const
 void
 BucketApplicator::advance()
 {
-    soci::transaction sqlTx(mDb.getSession());
-    while (mBucketIter)
+    LedgerState ls(mApp.getLedgerStateRoot(), false);
+    for (; mBucketIter; ++mBucketIter)
     {
-        LedgerHeader lh;
-        LedgerDelta delta(lh, mDb, false);
-
-        auto const& entry = *mBucketIter;
-        if (entry.type() == LIVEENTRY)
+        if ((*mBucketIter).type() == LIVEENTRY)
         {
-            EntryFrame::pointer ep = EntryFrame::FromXDR(entry.liveEntry());
-            ep->storeAddOrChange(delta, mDb);
+            auto const& bucketEntry = (*mBucketIter).liveEntry();
+            auto key = LedgerEntryKey(bucketEntry);
+            auto entry = ls.load(key);
+            if (entry)
+            {
+                entry.current() = bucketEntry;
+            }
+            else
+            {
+                ls.create(bucketEntry);
+            }
         }
         else
         {
-            EntryFrame::storeDelete(delta, mDb, entry.deadEntry());
+            auto entry = ls.load((*mBucketIter).deadEntry());
+            if (entry)
+            {
+                entry.erase();
+            }
         }
-        ++mBucketIter;
-        // No-op, just to avoid needless rollback.
-        delta.commit();
         if ((++mSize & 0xff) == 0xff)
         {
             break;
         }
     }
-    sqlTx.commit();
-    mDb.clearPreparedStatementCache();
+    ls.commit();
+    mApp.getDatabase().clearPreparedStatementCache();
```

Comment:
btw, should we still do these, or those "clear" statements should be done from within `LedgerStateRoot` ?

Reply:
This should be part of `LedgerStateRoot`, fixed.